File: android/src/main/kotlin/com/bugreport/PerfTracker.kt
package com.bugreport

/**
 * Performance tracing — mirrors JS BR.perf.start/end exactly.
 */
class PerfTracker(private val onLog: (Int, String, String, String, Map<String, Any?>?) -> Unit) {

    private val traces = mutableMapOf<String, Long>()
    private val DEFAULT_THRESHOLD = 3000L

    fun start(name: String) {
        traces[name] = System.currentTimeMillis()
    }

    fun end(name: String, thresholdMs: Long = DEFAULT_THRESHOLD): Long {
        val start = traces.remove(name) ?: return -1
        val duration = System.currentTimeMillis() - start

        val level = if (duration > thresholdMs) LogLevel.WARN.value else LogLevel.INFO.value
        val tag = "perf:$name"
        val msg = "${duration}ms" + if (duration > thresholdMs) " (threshold: ${thresholdMs}ms)" else ""

        val extra = mapOf<String, Any?>(
            "duration" to duration,
            "threshold" to thresholdMs,
            "name" to name
        )
        onLog(level, LogCategory.PERF.value, tag, msg, extra)
        return duration
    }

    fun mark(name: String, ms: Long) {
        val level = if (ms > DEFAULT_THRESHOLD) LogLevel.WARN.value else LogLevel.INFO.value
        onLog(level, LogCategory.PERF.value, "perf:$name", "${ms}ms", mapOf("duration" to ms))
    }

    fun clear() {
        traces.clear()
    }
}

File: ios/Sources/BugReport/PerfTracker.swift
import Foundation

/// Performance tracing — mirrors JS BR.perf.start/end and Kotlin PerfTracker.
public class PerfTracker {
    private var traces = [String: Int64]()
    private let defaultThreshold: Int64 = 3000
    private let onLog: (Int, String, String, String, [String: Any]?) -> Void

    public init(onLog: @escaping (Int, String, String, String, [String: Any]?) -> Void) {
        self.onLog = onLog
    }

    public func start(_ name: String) {
        traces[name] = Int64(Date().timeIntervalSince1970 * 1000)
    }

    @discardableResult
    public func end(_ name: String, thresholdMs: Int64 = 3000) -> Int64 {
        guard let start = traces.removeValue(forKey: name) else { return -1 }
        let duration = Int64(Date().timeIntervalSince1970 * 1000) - start

        let level = duration > thresholdMs ? LogLevel.warn.rawValue : LogLevel.info.rawValue
        let tag = "perf:\(name)"
        let msg = "\(duration)ms" + (duration > thresholdMs ? " (threshold: \(thresholdMs)ms)" : "")
        let extra: [String: Any] = ["duration": duration, "threshold": thresholdMs, "name": name]

        onLog(level, LogCategory.PERF.rawValue, tag, msg, extra)
        return duration
    }

    public func mark(_ name: String, ms: Int64) {
        let level = ms > defaultThreshold ? LogLevel.warn.rawValue : LogLevel.info.rawValue
        onLog(level, LogCategory.PERF.rawValue, "perf:\(name)", "\(ms)ms", ["duration": ms])
    }

    public func clear() {
        traces.removeAll()
    }
}
